### PR TITLE
Search: Show indexing progress in Performance Settings and Customizer

### LINKED ISSUE #15383: Search: Initial time to start indexing is too long and a bad UX
I spun up a jurassic.ninja site that was pre-populated with content and as a part of the set up I immediately purchased a search plan. This is just about the worst case for not allowing Jetpack to sync any data before a user has purchased search and is starting to run searches.

After the purchase I went to the customizer. The search customization works fine, but:
- there are no filters displayed because there are no results
- there are no results no matter what I search for

Looking on the backend I can see that the ES index was not populated for about 5 minutes (we have a 5 minute deduping queue for bulk jobs). Once it was populated the searches I ran before still were not showing any results because they had been cached in the browser. After reloading the browser they still were not shown because we have a five minute timeout on the search query caching if the hash of the search is identical to a previous one.

End result, the user can currently have to wait 10 minutes from after they purchase to when they can first see search results.

Ideas for changes:
- Change our messaging at the end of the purchase flow
- Reduce the query caching from 5 min to 2 minutes
- Reduce the dedupe queue or don't dedupe at all? I have some concerns with this, but maybe we can do it selectively...

cc @keoshi

### PR DESCRIPTION
**Pull request by jsnmoon**
Fixes #15383.

Screen Shot 2020-04-17 at 3 23 10 PM

Screen_Shot_2020-04-21_at_4_47_48_PM

#### Changes proposed in this Pull Request:
* Shows Jetpack Sync progress in Performance Settings (`/wp-admin/admin.php?page=jetpack#/performance`).
* Shows a notice in the Jetpack Search section of the customizer if sync hasn't yet finished.
* Adds a local API route returning the current Jetpack Sync status (`jetpack/v4/sync-progress`).
* Adds sync-progress state handling in Redux.
* Ports a ProgressBar component from Calypso.
* Adds Webpack configuration for a separate Jetpack Search Customizer JS bundle.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This change adds a progress bar inside the Performance Settings page to indicate indexing progress for the customer's site.
* This change adds a notice to the Jetpack Search section of the customizer to indicate indexing progress for the customer's site.

#### Testing instructions:
* Apply this patch to your Jetpack installation.
* Ensure that you have a Jetpack Search product subscription for your site.
* Navigate to the Performance Settings page for Jetpack (`/wp-admin/admin.php?page=jetpack#/performance`). Ensure that you can see a progress bar indicating search indexing status.

#### Proposed changelog entry for your changes:
* Indicate indexing progress to Jetpack Search customers in Performance Settings and in the Customizer.

### CODE REVIEW

**Review by AnnaMag — COMMENTED** [review]
I tested on a 1-tier site, thus the bar was shown complete immediately. 

After removing Search, the progress bar remains in Settings. The toggles seem to be removed/overwritten.
toggle

File: _inc/client/components/progress-bar/style.scss [hunk 1]
```diff
@@ -0,0 +1,60 @@
+@import '../../scss/variables/colors';
+
+.progress-bar {
+	width: 100%;
```
Comment (AnnaMag) [inline _inc/client/components/progress-bar/style.scss:4]: Currently the bar covers ~75% of the section width on complete indexing. Would it be a good idea to scale it to span the length of the space?

File: _inc/client/performance/search.scss [hunk 2]
```diff
@@ -0,0 +1,12 @@
+@import '../scss/variables/colors';
+
+.jp-form-search-setting-progress {
+	margin-bottom: 1em;
+
+	.progress-bar {
+		width: 75%;
```
Comment (AnnaMag) [inline _inc/client/performance/search.scss:7]: This looks a bit short to represent complete indexing.

Comment (jsnmoon) [reply _inc/client/performance/search.scss:7]: It does look a bit awkward in its current state. @keoshi, do you have any thoughts on this?

Currently, the progress bar looks like this in wp-admin:
Screen Shot 2020-04-17 at 3 23 10 PM

The design mock looks like this:
image

File: _inc/client/performance/search.jsx [hunk 3]
```diff
@@ -141,16 +163,20 @@ function Search( props ) {
...
-		isLoading: isFetchingSitePurchases( state ),
-		hasActiveSearchPurchase: selectHasActiveSearchPurchase( state ),
-		isBusinessPlan: 'is-business-plan' === planClass,
-		failedToEnableSearch:
-			! isSettingActivated( state, 'search' ) &&
-			! isUpdatingSetting( state, 'search' ) &&
-			false === hasUpdatedSetting( state, 'search' ),
-	};
-} )( withModuleSettingsFormHelpers( Search ) );
+export default connect(
```
Comment (AnnaMag) [inline _inc/client/performance/search.jsx:168]: Nice clean up!

**Review by gibrown — DISMISSED** [review]
This is working for me and looks good.

There are some issues with the way it is determining whether everything is sync-ed. We should really not only rely on the sync progress, but also on the indexing progress. And/Or we need to make sure that sync and indexing progress are more tightly tied together.

I think this is still a step in the right direction though so we should merge it. I had thought that @mdbitz and @roccotripaldi had a wp.com endpoint for the sync status, but this seems to be using the main sync module and not looking at the cache. I think we need to figure out how to get on the same page here, but this lays something good to build on.

**Review by AnnaMag — COMMENTED** [review]
Code LGTM and the bar is correctly displayed now. 

The status in the customizer does not change for me though and stays at `Your site is currently being indexed for search… ` even after the indexing is marked as complete in the Performance section. I do not see request firing for status update :thinking:

**Comment by jsnmoon** [comment]
> The status in the customizer does not change for me though and stays at `Your site is currently being indexed for search… ` even after the indexing is marked as complete in the Performance section.

Ah, nice catch! There was a typo in the boolean check for sync progress completion.

**Comment by mdbitz** [comment]
I've got to review this in full on Monday however to Greg's comments.

> There are some issues with the way it is determining whether everything is sync-ed. We should really not only rely on the sync progress, but also on the indexing progress. And/Or we need to make sure that sync and indexing progress are more tightly tied together.

This currently will only tell you the status of the Full Sync, it will not give us the status of the index which only triggers after the full sync completes. Stating "Your site search is fully operational" based on the sync status is going to lead to potential support issues as the index may not be fully built.

> I think this is still a step in the right direction though so we should merge it. I had thought that @mdbitz and @roccotripaldi had a wp.com endpoint for the sync status, but this seems to be using the main sync module and not looking at the cache. I think we need to figure out how to get on the same page here, but this lays something good to build on.

We were using a simple ajax response as defined in the [debug-functions](https://github.com/Automattic/jetpack/blob/3c23774c2fd7eb61b01e2fa982b672f219609cfa/_inc/lib/debugger/debug-functions.php#L160). It would make sense to align and not duplicate behavior across a REST API and ajax handler.

Also it should be discussed expectations for when this should be shown. I think it would only make sense while a sync is in progress. There are many scenarios when the cache site can get out of index and so saying all is well between full syncs is misleading.

**Comment by jsnmoon** [comment]
Closing this for now, further discussion can be found at p1HpG7-94Z-p2.

**Pull request closed without merging**